### ISSUE TITLE
Calling simulate() runs into conflicts if package 'lme4' is loaded

### ISSUE THREAD

**Issue by mbojan** [statnet/ergm#130]
This has been reported on Statnet Help here http://mailman13.u.washington.edu/pipermail/statnet_help/2019/002891.html in the context of ergm.ego but the problem is with ergm. It has also been reported here https://github.com/leifeld/btergm/issues/13 It can be reproduced by loading both 'ergm' and 'lme4' (in that order)

```r
library(ergm)
library(lme4)  # Conflict is reported by R (>= 3.6.0)
# Loading required package: Matrix
# Registered S3 method overwritten by 'lme4':
#  method           from
#  simulate.formula ergm
g.sim <- simulate(network(16) ~ edges + mutual, coef=c(0, 0))
# Error in weights(object) : argument "object" is missing, with no default
traceback()
# 4: weights(object)
# 3: .simulateFun(formula = object, nsim = nsim, seed = seed, family = family, 
#        weights = weights, offset = offset, ...)
# 2: simulate.formula(network(16) ~ edges + mutual, coef = c(0, 0))
# 1: simulate(network(16) ~ edges + mutual, coef = c(0, 0))
```

The problem is that lme4 has a non-exported method `lme4:::simulate.formula()` conflicting with `ergm::simulate.formula()`. The generic dispatches on the method which was loaded last, so from lme4.

I don't think there is a obvious way around this, apart from renaming functions in either ergm or lme4. I'm researching this now.

BUT, we already have `simulate_formula()`. Is the plan to perhaps deprecate `simulate.formula()` in favor of the underscored version, @krivit ? ~~Overloading the formula method to redirect the dispatch based on the LHS seems rather risky.~~

EDIT: Perhaps it is actually not that bad, but it will not solve the issue.

**Comment by mbojan** [comment]
Perhaps:

1. Warn about the conflict in the documentation. We could even add a custom warning to `.onLoad()` that screams at the user that some things will break if `"lme4" %in% loadedNamespaces()`.

**Comment by mbojan** [comment]
2. Deprecate `simulate.formula()` in favor of `simulate_formula()`.

**Comment by martinamorris** [comment]
@mbojan do you want to submit a pull request for this?

**Comment by krivit** [comment]
I could have sworn that I had replied to this ticket. (Is it duplicated? Is there a `summary.formula` version of it?) AFAIK, we first used `simulate.formula()` before they did, and our use is no less legitimate than theirs, so I am not ready to yield that syntax.

If we do yield, I am strongly opposed to `simulate_formula`, since it doesn't actually buy us anything while being unnecessarily verbose. We might want to use something like `rERGM()` or `rergm()` instead.

**Comment by martinamorris** [comment]
In principle I kind of like the idea of `rergm()` as it uses standard R naming conventions.  But (1) we have a long history of using the `simulate` language that I'm reluctant to abandon and (2) we don't have the corresponding `dergm()` and `pergm()` that this convention implies.

I thought the point of S3 methods was that they automagically identified the class of object and acted accordingly.  Is this a problem that `lme4` should fix?

**Comment by krivit** [comment]
`dergm()` is meaningful for calculating either likelihood or probability up to a constant. Nor would this be the only package that doesn't implement the full set of functions. For example `rmvtnorm` only implements `rmvnorm()` and `dmvnorm()`.

The problem is that both we and `lme4` want to dispatch for `formula`, but while we have a mechanism that further narrows things down using the class of the LHS, that's not the usual way things are done.

**Comment by martinamorris** [comment]
Is it correct to say that this general problem has the officially recommended solution of using the syntax package::function?  If so, can we adopt this internally, and recommend it to our users?

**Comment by krivit** [comment]
I guess if we re-exported `simulate.formula`, someone could use `ergm::simulate.formula`. I believe there might also be mechanisms for detecting if `lme4` is loaded after `ergm` and warning the user.

**Comment by leifeld** [comment]
The `lme4` folks have now exported `simulate.formula` (in addition to exporting it as a method), as per [this](https://github.com/lme4/lme4/issues/566) thread. If you did the same, this problem could be avoided at least in third-party packages because people could use `ergm::simulate.formula` or `lme4::simulate.formula` explicitly to distinguish between the two. Would you consider that?

**Comment by martinamorris** [comment]
This seems like a good idea.  @krivit any reason not to?

**Comment by krivit** [comment]
@leifeld , thanks for raising the issue with everyone, and thanks for taking the initiative with the PR, @mbojan . Before we do that, I want to see if we can make it seamless to the end-user by implementing LHS-dispatching in `lme4`, since they expect one-sided formulas, which we are discussing in lme4/lme4#566.

**Comment by krivit** [comment]
Tagging this as a bug, to distinguish it from issues that would be nice to have resolved by 3.11, but not critical. I think @bbolker and I are close to a solution.

**Comment by martinamorris** [comment]
Say hi to @bbolker for me :)
